### ISSUE TITLE
apicall hook `snprintf` function and variants

### ISSUE THREAD

**Issue by mr-tz** [mandiant/flare-floss#318]
Some decoding routines rely on string formatting routines such as `snprintf`. Example: `9dab...106c`

**Comment by samrath-sudesh-acharya** [comment]
@mr-tz can I take up this issue and could you guide me where to find example you mentioned

**Comment by mr-tz** [comment]
Of course, unfortunately, I don't have the sample handy anymore, but we can easily create a test C program that uses `snprintf` etc. to test API hooks we create. Can you go ahead with that or do you want me to provide more details?

**Comment by samrath-sudesh-acharya** [comment]
Ok, I will make the test C program and start working on this

**Comment by samrath-sudesh-acharya** [comment]
@mr-tz Sorry for the delay. I was using this C code to test the hook

```py
#include <stdio.h>
 
void print_string(char* buffer, int value) {
 
    snprintf(buffer, 100, "The value is: %d", value);
}
 
int main() {
    char buffer[100];
    int value = 42;
 
    print_string(buffer, value);
 
    printf("%s\n", buffer);
 
    return 0;
}
``` 

But the hook wasn't getting triggered. Is there something wrong I am doing ?

Here is [my code](https://github.com/samrath-sudesh-acharya/flare-floss/blob/403ccf77ef0ec983d0bd9cfd65b1d6a9582124b1/floss/api_hooks.py#L258)

**Comment by mr-tz** [comment]
That looks good. Does vivisect recognize the `snprintf` call correctly? If not you could link it dynamically.

**Comment by samrath-sudesh-acharya** [comment]
Ok yes I will check on that

**Comment by samrath-sudesh-acharya** [comment]
@mr-tz I have gone through the documentation of Vivisect and tried to play around with it. From how much I observe the the only way to find the `snprintf` call is to parse opcode from the basic block in the function and find out which set of operations is likely to do a snprintf call. I can't figure out how to find what set assembly-level code will likely resemble the call we are targeting. Could you guide me through or is my logic behind the implementation wrong?

**Comment by mr-tz** [comment]
Did you try dynamic linking?

**Comment by mr-tz** [comment]
For the dynamic linking I was referring to how the test binary is build. In the emulator hook it is fine to just reproduce the snprintf functionality in Python (no ctypes needed).

**Comment by Chee-Tzu** [comment]
@mr-tz Do you mean something like LD_PRELOAD'ing when you mean by dynamic linking?

**Comment by Chee-Tzu** [comment]
@mr-tz Sorry, I can't seem to get the hooking working. Can I get some guidance?
By dynamic linking, I hope being able to detect snprintf and its args, right?

**Comment by mr-tz** [comment]
Yes exactly, what issues are you encountering?

**Comment by codesensei-tushar** [comment]
Hi @mr-tz , I'm interested in implementing snprintf hook support for FLOSS.
Looking at the existing hooks in api_hooks.py, I see there's already a disabled PrintfHook with this TODO:

```python
# TODO disabled for now as incomplete (need to implement string format) and could result in FP strings as is
```
For snprintf, since it constructs strings dynamically using format specifiers, we'd need to parse the format string and substitute the argument values from the emulator state.

Is manually parsing and formatting each specifier the expected approach here? Or is there a simpler way to handle this that I might be missing?

Happy to work on a PR if you can confirm the direction.

**Comment by williballenthin** [comment]
> For snprintf, since it constructs strings dynamically using format specifiers, we'd need to parse the format string and substitute the argument values from the emulator state.

correct

> Is manually parsing and formatting each specifier the expected approach here? Or is there a simpler way to handle this that I might be missing?

this is reasonable, but we don't want to maintain a lot of complex code. so if you can find a simple solution for 80% of the scenarios (%d, %s, %f, etc.) then thats sufficient. alternatively, maybe there's a standard library function we can use for this.
